src/mono/mono/mini/jit-icalls.c

```diff
@@ -94,7 +95,10 @@ ldvirtfn_internal (MonoObject *obj, MonoMethod *method, gboolean gshared)
 
 	/* An rgctx wrapper is added by the trampolines no need to do it here */
 
-	return mono_ldftn (res);
+	addr =  mono_ldftn (res);
+	if (m_class_is_valuetype (res->klass) && !m_class_is_valuetype (method->klass))
+		addr = mini_add_method_trampoline (method, addr, FALSE, TRUE);
```

Comment:
I'm not sure if it makes a difference, but we should pass `res` here and not the original virtual `method`

Reply:
make sense